### PR TITLE
Topological divergence

### PR DESCRIPTION
**Pull request by ahdr3w**
Add TopologicalDivergence estimator based on the method proposed in [arXiv:2504.10063](https://arxiv.org/abs/2504.10063)

### CODE REVIEW

File: src/lm_polygraph/defaults/register_default_stat_calculators.py [hunk 1]
```diff
@@ -148,6 +148,30 @@ def _register(
...
+                "few_shot_split": "train",
+                "train_split": "train",
+                "load_from_disk": False,
+                "subsample_train_dataset": 100,
+                "n_shot": 0,
+                "batch_size": 1,
+                "seed": [1],
+                "size": None,
+            },
+        )
```
Comment (ArtemVazh) [inline src/lm_polygraph/defaults/register_default_stat_calculators.py:174]: Please move all these parameters into the configs and also specify this calculator there. Additionally, it might be much better to integrate the initialization of this calculator with `TrainingStatisticExtractionCalculator` to avoid multiple loadings of the training dataset.

For reference, we combine `TrainingStatisticExtractionCalculator` with other calculators for supervised methods in another PR. I think this case should be handled in a similar way.

calculator: https://github.com/IINemo/lm-polygraph/blob/supervised_methods/src/lm_polygraph/stat_calculators/statistic_extraction.py

initialization: https://github.com/IINemo/lm-polygraph/blob/supervised_methods/src/lm_polygraph/defaults/stat_calculator_builders/default_TrainingStatisticExtractionCalculator.py

config: https://github.com/IINemo/lm-polygraph/blob/supervised_methods/examples/configs/stat_calculators/default_calculators.yaml

File: src/lm_polygraph/defaults/stat_calculator_builders/default_TrainMTopDivCalculator.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
+import os
```
Comment (ArtemVazh) [inline src/lm_polygraph/defaults/stat_calculator_builders/default_TrainMTopDivCalculator.py:1]: To avoid multiple loadings of the training dataset, it might be much better to integrate the initialization of this calculator with `TrainingStatisticExtractionCalculator`

File: src/lm_polygraph/stat_calculators/train_mtopdiv.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+            heads = heads[: self.max_heads]
+            return {"topological_divergence_heads": heads}
+
+        train_dataset = self.load_train_dataset_fn()
+
+        mtopdivs = []
+        labels = []
+
+        greedy_calc = GreedyProbsCalculator(False, False)
+        attn_forward_pass_calc = AttentionForwardPassCalculator()
```
Comment (ArtemVazh) [inline src/lm_polygraph/stat_calculators/train_mtopdiv.py:81]: This should be specified in `__init__`

File: src/lm_polygraph/stat_calculators/train_mtopdiv.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        generation_metric = RougeMetric("rougeL")
```
Comment (ArtemVazh) [inline src/lm_polygraph/stat_calculators/train_mtopdiv.py:82]: Since `rougeL` is a questionable metric for UQ [1, 2, 3], it would be better to treat it as a hyperparameter and retain it only as a baseline option.

[1] D. Janiak, J. Binkowski, A. Sawczyn, B. Gabrys, R. Shwartz-Ziv, and T. Kajdanowicz, The Illusion of Progress: Re-evaluating Hallucination Detection in LLMs. 2025.  https://arxiv.org/abs/2508.08285 \
[2] Santilli, A. et al. Revisiting Uncertainty Quantification Evaluation in Language Models: Spurious Interactions with Response Length Bias Results. https://doi.org/10.48550/arXiv.2504.13677 (2025).
[3] Ielanskyi, M., Schweighofer, K., Aichberger, L. & Hochreiter, S. Addressing pitfalls in the evaluation of uncertainty estimation methods for natural language generation. (2025).

File: src/lm_polygraph/stat_calculators/train_mtopdiv.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+            mtopdivs.append(
+                get_mtopdivs(
+                    heads,
+                    length_responses,
+                    attn_weights_batch,
+                    n_jobs=self.n_jobs,
+                )
+            )
+            labels.append(
+                generation_metric(
```
Comment (ArtemVazh) [inline src/lm_polygraph/stat_calculators/train_mtopdiv.py:111]: Similarly, it may be combined with other calculators in the `TrainingStatisticExtractionCalculator`

https://github.com/IINemo/lm-polygraph/blob/supervised_methods/src/lm_polygraph/defaults/stat_calculator_builders/default_TrainingStatisticExtractionCalculator.py#L68

File: src/lm_polygraph/stat_calculators/train_mtopdiv.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+                    stats=stats,
+                    target_texts=target,
+                )
+            )
+
+        mtopdivs = np.concatenate(mtopdivs, axis=0)
+
+        labels = np.concatenate(labels)
+        labels = ~(labels > 0.3)
```
Comment (ArtemVazh) [inline src/lm_polygraph/stat_calculators/train_mtopdiv.py:120]: This hyperparameter should be specified in `__init__` and in configs

File: src/lm_polygraph/utils/dataset.py [hunk 7]
```diff
@@ -50,8 +50,8 @@ def select(self, indices: List[int]):
         Parameters:
             indices (List[int]): indices to left in the dataset.Must have the same length as input texts.
         """
-        self.x = [self.x[i] for i in indices]
-        self.y = [self.y[i] for i in indices]
+        self.x = [self.x[int(i)] for i in indices]
+        self.y = [self.y[int(i)] for i in indices]
```
Comment (ArtemVazh) [inline src/lm_polygraph/utils/dataset.py:54]: This change is unclear

File: src/lm_polygraph/stat_calculators/train_mtopdiv.py [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+
+        heads = self.select_heads(mtopdivs, labels)
+        heads = np.unravel_index(heads, (num_layers, num_heads))
+        heads = np.stack(heads, axis=1)
+
+        save_model_heads(self.cache_path, name_or_path, heads.tolist())
```
Comment (ArtemVazh) [inline src/lm_polygraph/stat_calculators/train_mtopdiv.py:126]: It should be configurable, since saving all statistics is not always required

File: test/test_estimators.py [hunk 9]
```diff
@@ -252,6 +253,28 @@ def test_attentionscore(model):
     assert isinstance(ue.uncertainty, float)
 
 
+def test_topological_divergence_fixed_heads(model):
+    estimator = TopologicalDivergence(
+        heads=[[0, 0], [0, 1], [1, 0], [1, 1]],
+        n_jobs=-1,
+    )
+    ue = estimate_uncertainty(model, estimator, INPUT)
+    get_reusable_executor().shutdown(wait=True)
```
Comment (ArtemVazh) [inline test/test_estimators.py:262]: The usage of `get_reusable_executor` is unclear

File: test/test_estimators.py [hunk 10]
```diff
@@ -252,6 +253,28 @@ def test_attentionscore(model):
...
+    assert isinstance(ue.uncertainty, float)
+
+
+def test_topological_divergence_select_heads(model):
+    estimator = TopologicalDivergence(n_jobs=-1)
+    ue = estimate_uncertainty(model, estimator, INPUT)
+    get_reusable_executor().shutdown(wait=True)
```
Comment (ArtemVazh) [inline test/test_estimators.py:269]: Can you clarify how this would work without a training dataset?

**Review by ArtemVazh — CHANGES_REQUESTED** [review]
